### PR TITLE
ARROW-8065: [C++][Dataset] Refactor ScanOptions and Fragment relation

### PR DESCRIPTION
**Pull request by fsaintjacques**
This is the first part of a refactor to make Fragment accessible without a Scan operation instance. This is a breaking change. It introduces the concept of a physical schema and read schema, these concepts are analogous to Avro writer and reader schema.

- Move ScanOptions at Fragment::Scan instead of a property
- Refactor Dataset::GetFragments(ScanOptions) to Dataset::GetFragments(Expression)
- Add Fragment::ReadPhysicalSchema
- Normalize Python's {Dataset,Fragment,Scanner}.{scan,to_table,to_batches}()

### CODE REVIEW

File: python/pyarrow/_dataset.pyx
Comment (jorisvandenbossche) [inline python/pyarrow/_dataset.pyx:509]: When using Fragment.scan, it uses the Fragment's physical schema for the resulting table? (since the Fragment is not aware of the dataset "read" schema?) 
If so, we should note that here in the docstring I think

File: python/pyarrow/tests/test_dataset.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_dataset.py:677]: Can you also test without passing the dataset's schema, and assert that the column_names are [f1, f2] ?

File: python/pyarrow/tests/test_dataset.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_dataset.py:686]: Keep this but where the columns selection is passe to `to_table` ?

**Review by jorisvandenbossche — COMMENTED** [review]
Looks good! Browsed through the C++ code, looked a bit more in detail at the python code. 

The skipped python tests are because the reconstruction of a fragment is not fully working yet? (didn't fully understand the relation to ARROW-8318)

Just to make sure I correctly understood it:

- A Fragment has only a physical schema, and is not aware of the dataset schema
- But when scanning the Fragment (or through `to_table`), you can still specify this dataset schema (so scanning dataset vs fragments can still give the same result, i.e. basically `dataset.to_table()` and `pa.concat([f.to_table(schema=dataset.schema) for f in dataset.get_fragments()])`)

File: cpp/src/arrow/dataset/file_base.cc
Comment (bkietz) [inline cpp/src/arrow/dataset/file_base.cc:217]:
```suggestion
    const WritePlan& plan,
```

Instead, `scan_options` should be derived from `plan.{schema,partitioning}` to project out columns implicit in the partitioning

File: cpp/src/arrow/dataset/scanner_internal.h
Comment (bkietz) [inline cpp/src/arrow/dataset/scanner_internal.h:64]: Slight tangent: why aren't `*_internal.h` excluded in lint_cpp_cli.py? It seems they should be

File: python/pyarrow/_dataset.pyx
Comment (bkietz) [inline python/pyarrow/_dataset.pyx:125]: This coercion is common, should we have `_unwrap_expression_default_true()`?

File: python/pyarrow/_dataset.pyx
Comment (bkietz) [inline python/pyarrow/_dataset.pyx:115]:
```suggestion
            Return fragments matching the optional filter, using explicit
            partition expressions and/or embedded information like Parquet's
            statistics.
```

File: python/pyarrow/_dataset.pyx
Comment (bkietz) [inline python/pyarrow/_dataset.pyx:147]:
```suggestion
        It produces a stream of ScanTasks. Each task is meant to be a unit of work
        to be dispatched by the user; they are not executed automatically. This allows
        customization of local task scheduling and execution.
```

**Review by bkietz — CHANGES_REQUESTED** [review]
I'm impressed by how surgical this change was, looking great

File: cpp/src/arrow/dataset/dataset.cc
Comment (fsaintjacques) [reply cpp/src/arrow/dataset/dataset.cc:84]: A unit test triggered a segfault. I added this to move on to the next failure.

File: python/pyarrow/tests/test_dataset.py
Comment (fsaintjacques) [reply python/pyarrow/tests/test_dataset.py:686]: Via kwargs to `self._scanner` then forwarding kwargs to `Scanner.from_fragment`

File: cpp/src/arrow/dataset/dataset.h
Comment (fsaintjacques) [reply cpp/src/arrow/dataset/dataset.h:85]: Doesn't compile.

File: cpp/src/arrow/dataset/file_base.cc
Comment (fsaintjacques) [reply cpp/src/arrow/dataset/file_base.cc:217]: That's a task for [ARROW-8382](https://jira.apache.org/jira/projects/ARROW/issues/ARROW-8382). The writer shouldn't be aware of such issue, it takes a file source, a format, a RecordBatchReader (the dropping should happen internally here) and writes it­.

Since the Writer API has no bindings and not in used, I added this to ARROW-8382 so we can prioritise  ARROW-8062 (parquet dataset) and ARROW-8318 (fragment stored in dataset).

File: cpp/src/arrow/dataset/dataset.h
Comment (fsaintjacques) [reply cpp/src/arrow/dataset/dataset.h:113]: Ditto.

File: cpp/src/arrow/dataset/dataset.cc
Comment (fsaintjacques) [reply cpp/src/arrow/dataset/dataset.cc:36]: The effort to open a JIRA is equal to implementing it. I added it.

**Comment by kszucs** [comment]
@ursabot build

File: python/pyarrow/_dataset.pyx
Comment (fsaintjacques) [reply python/pyarrow/_dataset.pyx:509]: Yes, but you can still pass schema (which is not documented. I'll add this.

**Comment by fsaintjacques** [comment]
Addressed most comments and updated followup ticket with what's missing. PTAL and merge quickly so we can unblock the blocked tickets :)

**Review by jorisvandenbossche — APPROVED** [review]
Looks good to me, defer to @bkietz to approve the C++ side ;)

**Review by bkietz — APPROVED** [review]

**Pull request closed without merging**
